Adopt WTF::isPowerOfTwo() in more places

cdumez opened the pull request:
#### 4247950644a3cb9384ec66836b1f890e366c4082
<pre>
Adopt WTF::isPowerOfTwo() in more places
<a href="https://bugs.webkit.org/show_bug.cgi?id=292625">https://bugs.webkit.org/show_bug.cgi?id=292625</a>

Reviewed by Geoffrey Garen and Darin Adler.

Adopt WTF::isPowerOfTwo() in more places and get rid of duplicates.

* Source/JavaScriptCore/runtime/PropertyTable.cpp:
(JSC::PropertyTable::PropertyTable):
* Source/JavaScriptCore/runtime/PropertyTable.h:
(JSC::isPowerOf2): Deleted.
* Source/WebGPU/WGSL/AttributeValidator.cpp:
(WGSL::AttributeValidator::visit):
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::maximumMiplevelCount):
* Source/WebKit/Shared/SharedStringHashTableReadOnly.cpp:
(WebKit::SharedStringHashTableReadOnly::setSharedMemory):
(WebKit::isPowerOf2): Deleted.

Canonical link: <a href="https://commits.webkit.org/294639@main">https://commits.webkit.org/294639@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/101f8f2d6e538822e4decff103c7aee9ea365b48

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102507 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22176 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12490 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107668 "Hash 101f8f2d for PR 45018 does not build (failure)") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53144 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22484 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30682 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/107668 "Hash 101f8f2d for PR 45018 does not build (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34973 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105513 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17421 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92524 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/107668 "Hash 101f8f2d for PR 45018 does not build (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17258 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10553 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52501 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/95178 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87074 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10623 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110043 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101116 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29639 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21857 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86967 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30003 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88720 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86562 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22031 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31391 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9111 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23889 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29567 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34883 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/124750 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29378 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34613 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32701 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30937 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebGPU/WGSL/AttributeValidator.cpp:
```diff
@@ -451,10 +452,9 @@ void AttributeValidator::visit(AST::StructureMember& member)
             if (alignmentValue < 1)
                 error(attribute.span(), "@align value must be positive"_s);
-            else if (!isPowerOf2)
+            else if (!isPowerOfTwo(static_cast<uint64_t>(alignmentValue)))
```
I am a little surprised a type cast is needed. Should we be using a function more like “make unsigned” instead of a cast to a specific type?

cdumez replied:
Maybe I could update isPowerOfTwo() to allow taking signed types and convert them to signed types as needed.

cdumez replied:
The type cast is needed because `WTF::isPowerOfTwo()` takes a T, which is passed to `std::has_single_bit()` internally, which also takes a `T` but enforces that the T be a unsigned integer type.

You mention using a function to "make unsigned". One function I've aware of is `isPowerOfTwo(static_cast<std::make_unsigned_t<decltype(alignmentValue)>>(alignmentValue)`. This is verbose and not super readable. As I said above, maybe I could do this inside the implementation of `isPowerOfTwo()` though, to let call sites call with a signed type.

That said, it feels like if we wanted to handle signed types in `isPowerOfTwo()`, we'd return false when the value is negative before blindly casting to an unsigned type.

cdumez replied:
Or maybe this?
```
isPowerOfTwo<std::make_unsigned_t<decltype(alignmentValue)>>(alignmentValue)
```

darinadler replied:
I guess I’m thinking we’d make `unsigned_cast` or `make_unsigned_cast`, which would use `make_unsigned_t` in its implementation, but spell it `unsignedCast`. Like `byteCast` it would be a less powerful cast for a specific purpose, it would be `isPowerOfTwo(unsignedCast(alignmentValue))`.

cdumez replied:
Ok. I can look into introducing such as casting function and adopt it here.

cdumez replied:
Following up in https://github.com/WebKit/WebKit/pull/45095

darinadler commented on Source/JavaScriptCore/runtime/PropertyTable.h:
```diff
@@ -59,11 +59,6 @@ JS_EXPORT_PRIVATE extern PropertyTableStats* propertyTableStats;
 
-inline constexpr bool isPowerOf2(unsigned v)
-{
-    return hasOneBitSet(v);
```
Is the `hasOneBitSet` function still being used?

cdumez replied:
Yes, it is still being used a lot. I am planning to follow-up though since `std::has_single_bit()` exists in C++20. That said, we have the issue that `hasOneBitSet()` can be called with signed types while `std::has_single_bit()` cannot.

cdumez committed:
Adopt WTF::isPowerOfTwo() in more places
https://bugs.webkit.org/show_bug.cgi?id=292625

Reviewed by Geoffrey Garen and Darin Adler.

Adopt WTF::isPowerOfTwo() in more places and get rid of duplicates.

* Source/JavaScriptCore/runtime/PropertyTable.cpp:
(JSC::PropertyTable::PropertyTable):
* Source/JavaScriptCore/runtime/PropertyTable.h:
(JSC::isPowerOf2): Deleted.
* Source/WebGPU/WGSL/AttributeValidator.cpp:
(WGSL::AttributeValidator::visit):
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::maximumMiplevelCount):
* Source/WebKit/Shared/SharedStringHashTableReadOnly.cpp:
(WebKit::SharedStringHashTableReadOnly::setSharedMemory):
(WebKit::isPowerOf2): Deleted.

Canonical link: https://commits.webkit.org/294639@main
